lib/screens/v2/candidate_generator/widgets/reconstructed_alert.ex

```diff
@@ -79,22 +79,69 @@ defmodule Screens.V2.CandidateGenerator.Widgets.ReconstructedAlert do
     Enum.member?(@relevant_effects, effect)
   end
 
-  defp relevant_location?(%ReconstructedAlert{alert: alert} = reconstructed_alert) do
+  defp relevant_location?(%ReconstructedAlert{} = reconstructed_alert) do
     case BaseAlert.location(reconstructed_alert) do
       location when location in [:downstream, :upstream] ->
         true
 
       :inside ->
-        alert.effect != :delay or alert.severity > 3
+        relevant_inside_alert?(reconstructed_alert)
 
       location when location in [:boundary_upstream, :boundary_downstream] ->
-        alert.effect != :station_closure and (alert.effect != :delay or alert.severity > 3)
+        relevant_boundary_alert?(reconstructed_alert)
 
       _ ->
         false
     end
   end
 
+  defp relevant_inside_alert?(%ReconstructedAlert{alert: alert} = reconstructed_alert) do
+    alert.effect != :delay or
+      (alert.severity > 3 and relevant_direction?(reconstructed_alert))
+  end
+
+  defp relevant_boundary_alert?(%ReconstructedAlert{alert: alert} = reconstructed_alert) do
+    alert.effect != :station_closure and
+      (alert.effect != :delay or
+         (alert.severity > 3 and relevant_direction?(reconstructed_alert)))
+  end
+
+  # This function assumes that stop_sequences is ordered by direction north/east -> south/west.
+  # If the current station has only one stop sequence and the stop_id is the first or last entry in the stop_sequence,
+  # it is a terminal. Delay alerts heading in the direction of the station are not relevant.
+  defp relevant_direction?(
+         %ReconstructedAlert{
+           screen: %Screen{
+             app_params: %{reconstructed_alert_widget: %CurrentStopId{stop_id: stop_id}}
+           },
+           stop_sequences: [stop_sequence]
```

Comment:
Hmm, I guess this code wouldn't work if we ever have a terminal stop that's also a transfer stop (served by multiple lines). In that case, there would be more than one stop sequence in the list.

E.g., imagine if the blue line ended at Government Center instead of Bowdoin. Government Center would be a terminal stop for the blue line.
![image](https://user-images.githubusercontent.com/63608771/166512672-485d8b26-6401-4e19-a6df-014591e1aa18.png)

While that assumption is safe to make now, it might not be forever, and it would be hard to catch the incorrect behavior for that specific edge case in the future.

What do you think about doing this instead, as a more general approach?
```ex
defp relevant_direction?(... stop_sequences: stop_sequences ...) do
  first_stops = MapSet.new(stop_sequences, &List.first/1)
  last_stops = MapSet.new(stop_sequences, &List.last/1)

  relevant_direction_for_terminal =
    cond do
      ...
      stop_id in first_stops -> 0
      stop_id in last_stops -> 0
      ...
    end

  ...
end
```

Reply:
This was something I got caught up on during implementation, but I couldn't quite find the right solution. This is closer than I got, but I think it might be a little too generic. In this example (I know it's just a hypothetical but bear with me), BL trains from Gov't Center travel East and GL trains travel West. So they will always be filtering each other out. I think the biggest obstacle to this problem is not knowing which stop sequence belongs to which line. If we knew that, I think this approach would be great. I'll look into this, but do you know of a way to get this from V3?

Reply:
I think I see what you mean. I wonder if we could get around this issue by working with platform-level stop IDs in this check, rather than station IDs? That way, a delay alert on the BL only informing the BL platform would only match up with the stop sequence that contains BL platform IDs. (and vice versa for GL)

Related: do we get multiple stop sequences for any stop on the GL? I thought we got separate ones for the trunk and the different branches. If that's the case then this will also run into trouble for any terminal GL stop 😬